### PR TITLE
CONTRIB-8152: add guest access feature

### PR DESCRIPTION
**Pull request by univietw**
This is the code for the guest access feature described in CONTRIB-8152

### CODE REVIEW

**Comment by moodlebeuth** [comment]
+1

**Comment by jfederico** [comment]
Hey @univietw thanks for the PR. However, I can see that you are still working on it. and you are causing us a big overhead (with each forced push you do) on travis 
image

It is allright, you can keep working on it. I am just going to close this PR, you can reopen it once your code is clean.

A piece of advice, enable your own repo in travis, this way you will see if your changes are passing even before opening the PR.

Also, it is highly recommended to run the codechecker before hand, it is not a perfect linter, but it is the best we got.

Thanks again

**Comment by univietw** [comment]
The pipeline should succeed now. Would you mind reopening it, because I don't have the rights to do that. Otherwise I will create a new one

File: guestlink.php
Comment (jfederico) [inline guestlink.php:22]: I am sorry. While we appreciate the contribution. This is a no-no for us. You can keep the authorship, but you would need to give away the copyright in favour of  * @copyright 2010 onwards, Blindside Networks Inc

File: version.php
Comment (jfederico) [inline version.php:29]: You have to restore the version to what we have 2019101001

File: bbb_ajax.php
Comment (jfederico) [inline bbb_ajax.php:40]: Wht is the new delete used for?

File: lib.php
Comment (jfederico) [inline lib.php:1117]: Is it  my imagination or this POR includes some other customizations you have?

File: classes/locallib/config.php
Comment (jfederico) [inline classes/locallib/config.php:77]: This is unrelated to this PR

**Review by jfederico — CHANGES_REQUESTED** [review]

**Comment by jfederico** [comment]
This is pretty cool and I really thank you for your contribution. But we have to focus on one issue at the time. This seems to me that it covers several others. I am going to have to do a more detailed review. we may take some ideas from here, but as it is, I could not merge it.

File: db/upgrade.php
Comment (jfederico) [inline db/upgrade.php:246]: Still working on this review. This is brutal for massive databases. I am thinking that a better approach is to keep the empty and change the way they are implemented in the UI. Actually, base on this idea, the feature comes disabled by default which is what we want.

So, if the gueslinkid is empty, the feature is disabled. In the setting there should be only a checkbox that when ticked, it would open any other setting needed. On save, a new guestlinkid would be generated.

The UI only works is guestlink feature is enabled and properly set, so, also the view must be changed

File: guestlink.php
Comment (jfederico) [inline guestlink.php:58]: I am not found on this `custom_logo` idea. I understand that the idea is to differenciate the guestlink page from the rest to let the user know that this is for entering BBB, but most Universities/Colleges actually customize or ask customization to their theme. Why would we force them to do different if they have already their theme?

I think the guestlink page should be as any other moodle page. Perhaps like the Sign in page.

**Review by jfederico — COMMENTED** [review]
More thoughts

File: guestlink.php
Comment (jfederico) [reply guestlink.php:22]: Nevermind, I misunderstood this with licencing. It seems to be Ok as it is. You may need to sign a CLA though before we merge, but that is a different story.

I think I will need to do serious changes to UI, but in principle your implementation is outstanding. Thanks again.

File: version.php
Comment (univietw) [reply version.php]: This is fixed now

File: bbb_ajax.php
Comment (univietw) [reply bbb_ajax.php:40]: The delete is used for deleting the guest password

File: lib.php
Comment (univietw) [reply lib.php:1153]: The Copy is used for a small button for copying the link or guest password. I checked the code, there is nothing except for the guest access feature

File: classes/locallib/config.php
Comment (univietw) [reply classes/locallib/config.php:77]: The logo is for the guest access page. We did everything exactly like in greenlight. More on that above.

File: db/upgrade.php
Comment (univietw) [reply db/upgrade.php]: This is a good idea, I changed that in the updated version of the code.

File: guestlink.php
Comment (univietw) [reply guestlink.php:58]: We had a discussion about this ourselves. There are some things to consider:
a) it is not possible to use the standard moodle page since guests don't have access to any other content in moodle. So they can't see the navigation and so on.
b) not everyone has a signin page, since many sites are behind some SSO. Since this is the case we can't rely on any login page to be working. Even if there is a login page: The theme can change the way it is processed and there is no real way to ensure that a hook or something like that at least works on every moodle instance.
So we oriented on greenlight. And in greenlight you at least have the chance to change the logo, which is something I have seen on some other moodle plugins as well. Anyone can still change the CSS of this page, it's one of the most simplistic ones I have seen in my carreer and should just take some minutes.
If you have an idea how to solve the issue I am open for suggestions.

**Comment by goggo24** [comment]
We just pushed a newer version - hope you like it. Greetings from Uni Vienna.

**Comment by bwurst** [comment]
Hey there, I just found this pull request while searching for the same functionality. Tried it out immediately.

I wonder why this code calls the function bigbluebuttonbn_view_bbbsession_set() in line 81 of guestlink.php while this functions seems defined nowhere. 

Is commit 78223f7589e566fff708d10a2453fee2bac60e42 really your current code or did I miss something? It seems like the last change was in July 2020 according to git logs.

File: db/upgrade.php
Comment (bwurst) [reply db/upgrade.php]: With the current behavior, the link changes on every save button press. So if you publish a link to your external auditors, you must not change conference presets any more. That's a big issue as the moderator is not warned about that. Just changing a typo in room description will invalidate all links.

**Comment by bwurst** [comment]
Okay, when I replace line 81 in guestlink.php with 
\mod_bigbluebuttonbn\locallib\bigbluebutton::view_bbbsession_set($context, $bbbsession);
then the module works! 

First of all: Thank you all for this, we're really looking forward to use this. I made a german localization in my fork of this branch.

What I do not like: The public access link is changed whenever I hit Save in the meeting activity settings. So changing a typo in meeting description or changing an access rule for Moodle users will always invalidate the link. That's not good when you send the link e.g. to parents of pupils for a parents conference. I think the link should be static for a given meeting.

A great extension would be to use BBB's moderator approval. Thereforce, guest accesses must be identified with guest=true and the conference should get an option guestPolicy=(ALWAYS_ACCEPT || ASK_MODERATOR).

**Comment by bwurst** [comment]
I fixed both issues in my fork: https://github.com/bwurst/moodle-mod_bigbluebuttonbn/tree/CONTRIB-8152

**Comment by univietw** [comment]
Hi, @bwurst thanks for fixing the issues. We are very busy preparing the upcoming moodle-upgrade and I appreciate the help very much. Those two things were bugs which I seemed to have introduced while fixing the changes @jfederico had requested. If it's OK for you I would merge your branch into mine. So everything works fine.
@jfederico Is there anything else I could do to help merging the feature into master? The travis-ci seems to have crashed, in my branch I used, everything was green.

**Comment by bwurst** [comment]
Yes, please feel free to use my branch where suitable. I fixed the changing guestlinkid by reading it directly from the database. Don't know if it could or should be done more elegant.

The moderator approval is configurable in my branch, this alters the database again and introduces a new option. I handle accesses via guestlink as guest accesses and registered moodle users as non-guests.

I introduced the strings in german translation but I don't know how this is handled in general as the module does not have translation files in the code. I put my branch in production use last weekend and got no complaints so far. 

The usability of the guestlink UI is quite different from the rest of moodle (I cannot really describe, just feels different) so possibly someone should have a look and suggest an alternative UI.

**Comment by bwurst** [comment]
I tried to get clear about my concerns: What I dislike is that you set the password on the view page. This is not the place to change settings.
When you share the link with a code and then accidentially click on the reload-styled icon or the trashbin icon then you will never be able to restore the previous code. The action icons don't have text and in fact they miss a tool tip, so I'd be the first one to click around to see what's up there. 

Why didn't you set the guestlink password in the settings and allow free text entry, additional to a generate button?

Is there a reason that the code is numerical at all? Alphanumeric may be easier to memorize. I think all code logic is safe to use regular string password, given some security checks.

**Comment by univietw** [comment]
We mostly copied the feature from greenlight and wanted to keep it as close to there as possible.
That's why the code is numerical (as it is numerical in greenlight as well) It's the same reason why you can't recover it or set it yourself, as you can't do that in greenlight either.
The password should be visible to all people who have the right to, which is why we put it on the main screen. And since the password is there we thought it would be natural to be able to change it there as well. But you're right it would also make sense to change it only in the activity settings.
The only concern about that was that you would not be able to see the password right away, but the link itself isn't visible there either. I don't have an opinion which solution is the better one to be honest.
About the tooltip: There should be a tooltip, at least our production version has one. But it might be another error because of rebasing. I will have a look into that.

**Comment by brendanheywood** [comment]
We were about to embark on implementing this ourselves and then I discovered this pr. @univietw is this still in progress? Is there anything we can do to help get it over the line?

File: guestlink.php
Comment (brendanheywood) [reply guestlink.php:58]: The intention of the 'embedded' layout is for use inside of iframes and similar. I don't understand why a normal page layout can't be used, in particular the 'login' layout, or if that isn't good then 'base' or 'secure' should work. There are plenty of other similar page which are visible to the public, even if moodle guest mode is switched off.

**Comment by univietw** [comment]
hi @brendanheywood,
I started working on it again today. I am rebasing the code to the actual version and will test it on my local machine, then i will put the patch of @bwurst into it.
My hope is to include this feature before bigbluebuttonbn will be integrated into moodle core in moodle 3.11.

**Comment by brendanheywood** [comment]
ok thanks heaps @univietw - we'll happily test it even in a draft form so please keep us posted

**Comment by bwurst** [comment]
We are using the version that is current in my repository on https://github.com/bwurst/moodle-mod_bigbluebuttonbn.git branch CONTRIB-8152. 

It allowed us to set up face to face meetings with our parents using moderator approval (so one family at a time is in the meeting and others are in a queue until approved). That's a feature that the moodle module does not implement in any way and it eased up a lot here.

It would be great if that could be integrated into the official module and into moodle. You probably should rework the design of the guestlink page. It's heavily based on the look of greenlight and has some drawbacks when the conference name is long on a narrow display. When you integrate it into moodle, it should look more like moodle, not like greenlight. Users will probably not know greenlight at all.

**Comment by univietw** [comment]
hi @bwurst @brendanheywood I finished with my tasks (for now) and changed the following things:
1. The branch is now up to date with the master (hurray!)
2. I fixed the bug @bwurst found (the line in guestlink.php)
3. I did NOT integrate @bwurst changes (yet). I already had it in my branch but got confused and deleted the wrong one. I will reado it soon, it's not that hard (maybe today if I'm feeling for it)
4. The most important fix: I changed the guestlink page so it integrates into moodle, not into greenlight.
As I only have so many themes to test it with I would be very happy for anyone to give feedback if it works with their custom theme.

**Comment by brendanheywood** [comment]
thanks @univietw - we hopefully will start testing this next week

**Comment by ryprfpryr** [comment]
Hi there @univietw @bwurst 
Thank your for your contribution! We have been testing your branch CONTRIB-8152 and it works nicely!

I'd like to send you a pull request but I need a little help first. I would like to add a custom parameter to the room settings, specifically the "userdata-bbb_skip_check_audio", so users can enable/disable echo test for each BBB session from the Moodle's room settings.

I have tried by adding this exactly as if it where a Lock setting, but it has no effect. Could you please take a look at it? I would be glad to help adding this feature! Here's my branch: https://github.com/ryprfpryr/CONTRIB-8152_mod.git

Thanks!

**Comment by bwurst** [comment]
At first: I cannot help you as I do not have deeper knowledge about this code. Your branch is a complete reupload and contains one commit with all code. So nobody really can see which are your changes.

I'd not mix those pull requests together as I cannot see why this should depend on each other.

Please separate your changes (fork the master code to your github account, do your changes and upload them to your fork) and make a pull request against the master branch. The code in this pull request is not 100% up to date and should not be made more complex than it already is.

**Comment by ryprfpryr** [comment]
@bwurst sorry about that. 
We have pushed the "skip echo test" changes to the forked repo so you can take a look at the changes made: https://github.com/ryprfpryr/moodle-mod_bigbluebuttonbn/tree/CONTRIB-8152

**Comment by abias** [comment]
@ryprfpryr , I would like to second what @bwurst said: We should not mix up this feature / PR here with your suggested feature about skipping the echo test.

I would like to encourage you to create a separate PR just for this feature which does not build up the changes from this PR here but which is based on the latest release of this plugin instead.

This way, the plugin maintainers can review and merge your feature easily.

**Comment by abias** [comment]
Hi everyone, 

in collaboration with @univietw and Moodle partner @lernlink, I have reviewed this contribution over the last days:

* First, I have rebased the PR branch onto the latest upstream v2.4-stable branch to make the PR mergable again.
* Additionally, I have polished the PR to better fit into the existing code structure, look & feel and behaviour of mod_bigbluebuttonbn. During this work, I have especially added admin settings which make the guest access feature configurable. Please see the commit history for all details.
* Furthermore, I have added a feature which lets moderators see the guest access link not only in the Moodle activity but also in the BBB room chat. This way, moderators don't need to switch back to Moodle when they need to invite another participant from a running meeting.

As a result, I have just force-pushed the branch of the PR with my additions and changes.

@univietw and me are looking forward to your feedback to see what remains to be done before this feature could be integrated into the plugin.

@brendanheywood , it would be awesome if you at Catalyst could give the updated PR a test drive and could share your results here.

Cheers,
Alex

**Comment by abias** [comment]
PS: I noticed that all checks have failed for this PR now. This is most probably because of current problems with MariaDB in moodle-plugin-ci (see https://github.com/moodlehq/moodle-plugin-ci/pull/115 / https://tracker.moodle.org/browse/MDL-72131) and not a problem of this PR. As the repo does not have any other checks, especially on Postgres, configured, I can't really tell yet if there are any other remaining issues which would be raised by automated testing.

**Comment by brendanheywood** [comment]
thanks for the heads up @abias - I'll get @keevan to take a look, we've been actively testing this and working on cleaning up this feature too so it's likely we doubled up a bit and diverged somewhat, and I'm sorry we should have done more comms here. We'll get back to you

**Comment by jfederico** [comment]
Closing in favour of core implementation https://tracker.moodle.org/browse/MDL-74638 in Moodle 4.1

**Pull request closed without merging**
